src/qt/walletcontroller.cpp

```diff
@@ -147,23 +146,64 @@ void WalletController::removeAndDeleteWallet(WalletModel* wallet_model)
     delete wallet_model;
 }
 
+WalletControllerActivity::WalletControllerActivity(WalletController* wallet_controller, QWidget* parent_widget)
+    : QObject(wallet_controller)
+    , m_wallet_controller(wallet_controller)
+    , m_parent_widget(parent_widget)
+{
+}
 
-OpenWalletActivity::OpenWalletActivity(WalletController* wallet_controller, const std::string& name)
-    : m_wallet_controller(wallet_controller)
-    , m_name(name)
-{}
+WalletControllerActivity::~WalletControllerActivity()
+{
+    delete m_progress_dialog;
+}
 
-void OpenWalletActivity::open()
+void WalletControllerActivity::showProgressDialog(const QString& label_text)
 {
-    std::string error, warning;
-    std::unique_ptr<interfaces::Wallet> wallet = m_wallet_controller->m_node.loadWallet(m_name, error, warning);
-    if (!warning.empty()) {
-        Q_EMIT message(QMessageBox::Warning, QString::fromStdString(warning));
-    }
-    if (wallet) {
-        Q_EMIT opened(m_wallet_controller->getOrCreateWallet(std::move(wallet)));
-    } else {
-        Q_EMIT message(QMessageBox::Critical, QString::fromStdString(error));
+    m_progress_dialog = new QProgressDialog(m_parent_widget);
+
+    m_progress_dialog->setLabelText(label_text);
+    m_progress_dialog->setRange(0, 0);
+    m_progress_dialog->setCancelButton(nullptr);
+    m_progress_dialog->setWindowModality(Qt::ApplicationModal);
+    GUIUtil::PolishProgressDialog(m_progress_dialog);
+
+    connect(m_progress_dialog, &QObject::destroyed, [this] {
+        m_progress_dialog = nullptr;
+    });
+}
+
+OpenWalletActivity::OpenWalletActivity(WalletController* wallet_controller, QWidget* parent_widget)
+    : WalletControllerActivity(wallet_controller, parent_widget)
+{
+}
+
+void OpenWalletActivity::finish()
+{
+    m_progress_dialog->hide();
+
+    if (!m_error_message.empty()) {
+        QMessageBox::critical(m_parent_widget, tr("Open wallet failed"), QString::fromStdString(m_error_message));
+    } else if (!m_warning_message.empty()) {
+        QMessageBox::warning(m_parent_widget, tr("Open wallet warning"), QString::fromStdString(m_warning_message));
     }
+
+    if (m_wallet_model) Q_EMIT opened(m_wallet_model);
+
     Q_EMIT finished();
 }
+
+void OpenWalletActivity::open(const std::string& path)
+{
+    QString name = path.empty() ? QString("["+tr("default wallet")+"]") : QString::fromStdString(path);
+
+    showProgressDialog(tr("Opening Wallet <b>%1</b>...").arg(name.toHtmlEscaped()));
+
+    QTimer::singleShot(500, worker(), [this, path] {
```

Comment:
We can ditch these 500ms but think it is ok.

If we decide to have 0ms then we can only replace `QTimer::singleShot(0, ` with QMetaObject::invokeMethod after bumping Qt to 5.10 - see https://doc.qt.io/qt-5/qmetaobject.html#invokeMethod-5.

Reply:
What is the point to pause?
If there is no clear and obvious rationale, `QTimer::singleShot()` should be avoided, IMO.

Reply:
No point, but not harm, can make it 0. The point is to call the lambda in the `worker()` thread.

Reply:
> The point is to call the lambda in the `worker()` thread.

Did you consider [some other approaches](https://doc.qt.io/qt-5/qthread.html#details)?

Reply:
Yes, those require more QObject subclass,  signals and slots. IMO this approach is more clear.

Reply:
May I suggest to remove (make 0 ms) or comment 500 ms timeout?

Reply:
Sure, I can make 0ms and leave a comment.

Reply:
Is there a benefit of using 500ms? Seems racy...

Reply:
Not really, I'll make 0 then.